### PR TITLE
HIVE-24706: add the HiveHBaseTableInputFormatV2 to fix the compatible…

### PR DESCRIPTION
**Pull request by alexdongli0829**
### What changes were proposed in this pull request?

For HIVE-24706, the main issue here is the HiveHbaseTableInput format implements two version of InputFormat, which make the spark cannot get the correct version correctly, and this is indeed not very clear implementation.

So in this request, instead of directly extending TableInputFormatBase, I put it as a delegate which do the exactly the same as before, but avoid the confusing because the HbaseStorageHandler just need the old version InputFormat.

In the long term, I think hive should update the storage handler instead of keep mixing these two different API versions

### Why are the changes needed?

Its impacting the spark and hive compatible and reported by different uses in hive and spark

### Does this PR introduce _any_ user-facing change?
There is configuration parameter added hive.hbase.inputformat.v2, so maybe need update doc to keep the end user informed

### How was this patch tested?

create hbase table

```
echo "create 'students','account','address'" | sudo -u hbase hbase shell -n
echo "put 'students','student1','account:name','Alice'" |sudo -u hbase hbase shell -n
```

create hive table
```
hive -e "create external table test1 (key string, value string)
> stored by 'org.apache.hadoop.hive.hbase.HBaseStorageHandler'
> with serdeproperties ('hbase.columns.mapping' = ':key,account:name')
> tblproperties ('hbase.table.name' = 'students')"


SLF4J: Class path contains multiple SLF4J bindings.
Logging initialized using configuration in file:/etc/hive/conf.dist/hive-log4j2.properties Async: true
Hive Session ID = 05b4ec22-d15f-4614-9bc3-6c183e868728
OK
Time taken: 2.913 seconds
```

Spark test:

spark-sql --jars /usr/lib/hive/lib/hive-hbase-handler.jar,/usr/lib/hbase/hbase-common-2.4.4.jar,/usr/lib/hbase/hbase-client-2.4.4.jar,/usr/lib/hbase/lib/hbase-mapreduce-2.4.4.jar,/usr/lib/hbase/lib/shaded-clients/hbase-shaded-client-2.4.4.jar --conf spark.hive.hbase.inputformat.v2=true

```
spark-sql> select * from test1;

student1    Alice
```

Unit Test

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.hadoop.hive.hbase.TestHiveHBaseTableInputFormatV2
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.069 s - in org.apache.hadoop.hive.hbase.TestHiveHBaseTableInputFormatV2
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
```

### CODE REVIEW

**Comment by igkim89** [comment]
Thank you!
I was able to do read HBase-backed Hive tables in Spark 3.3.1 by applying this code.

**Comment by alexdongli0829** [comment]
Waiting for the review

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 1]
```diff
@@ -3505,6 +3505,11 @@ public static enum ConfVars {
     HIVE_HBASE_GENERATE_HFILES("hive.hbase.generatehfiles", false,
         "True when HBaseStorageHandler should generate hfiles instead of operate against the online table."),
     HIVE_HBASE_SNAPSHOT_NAME("hive.hbase.snapshot.name", null, "The HBase table snapshot name to use."),
+
+    HIVE_HBASE_INPUTFORMAT_V2("hive.hbase.inputformat.v2", false, "If enabled, the new " +
+            "version (V2) of input format that inherits only mapred version of InputFormat will be used as input format " +
```
Comment (aturoczy) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:3510]: I would use this message:
_The new version (V2) of the input format, which inherits only the mapred version of InputFormat, will be utilized as the default input format for reading the Hbase table. Currently, the old version (HiveHBaseTableInputFormat) is used by default, which inherits both the mapred and mapreduce versions of InputFormat._

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+      conn = ConnectionFactory.createConnection(HBaseConfiguration.create(jobConf));
+      delegate.initializeTableDelegate(conn, tableSplit.getTable());
+      delegate.setScan(HiveHBaseInputFormatUtil.getScan(jobConf));
+      recordReader = delegate.createRecordReader(tableSplit, tac);
+      try {
+        recordReader.initialize(tableSplit, tac);
+      } catch (InterruptedException e) {
+        delegate.closeTableDelegate(); // Free up the HTable connections
+        conn.close();
+        throw new IOException("Failed to initialize RecordReader", e);
```
Comment (aturoczy) [inline hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:97]: missing space and : 
(minor)

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+
+    final org.apache.hadoop.mapreduce.RecordReader<ImmutableBytesWritable, Result> recordReader;
+
+    Job job = new Job(jobConf);
+    TaskAttemptContext tac = ShimLoader.getHadoopShims().newTaskAttemptContext(job.getConfiguration(), reporter);
+
+    final Connection conn;
+
+    synchronized (HBASE_TABLE_MONITOR) {
+      conn = ConnectionFactory.createConnection(HBaseConfiguration.create(jobConf));
```
Comment (aturoczy) [inline hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:88]: This connection is not in try block and I do not see where this connection is closed. 
Only the catch contains close. Maybe just because of the synchronized, but hard to follow.

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+      String timestampColumn) {
+
+    IndexPredicateAnalyzer analyzer = new IndexPredicateAnalyzer();
+
+    // We can always do equality predicate. Just need to make sure we get appropriate
+    // BA representation of constant of filter condition.
+    // We can do other comparisons only if storage format in hbase is either binary
+    // or we are dealing with string types since there lexicographic ordering will suffice.
+    if (isKeyComparable) {
+      analyzer.addComparisonOp(keyColumnName, "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqual",
```
Comment (aturoczy) [inline hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:241]: I would put this UDF string in one stringbuilder or string if you use it at least twice.

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+          "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqualOrLessThan",
+          "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPLessThan",
+          "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPGreaterThan");
+    } else {
+      analyzer.addComparisonOp(keyColumnName, "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqual");
+    }
+
+    if (timestampColumn != null) {
+      analyzer.addComparisonOp(timestampColumn, "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqual",
+          "org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqualOrGreaterThan",
```
Comment (aturoczy) [inline hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:252]: Maybe wrong question but it is not a good idea to validate if this UDF's are presented? Maybe it is too defensive.

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+
+  @Override public InputSplit[] getSplits(final JobConf jobConf, final int numSplits) throws IOException {
+    synchronized (HBASE_TABLE_MONITOR) {
+      final UserGroupInformation ugi = UserGroupInformation.getCurrentUser();
+      if (ugi == null) {
+        return getSplitsInternal(jobConf, numSplits);
+      }
+
+      try {
+        return ugi.doAs(new PrivilegedExceptionAction<InputSplit[]>() {
```
Comment (aturoczy) [inline hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:269]: Why do you need doAs? As Hive user has all the right is should not need IMHO.

File: hbase-handler/src/test/org/apache/hadoop/hive/hbase/TestHiveHBaseTableInputFormatV2.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import static org.junit.Assert.assertTrue;
+
+
+/**
+ *  Test tot make sure the HiveHBaseTableInputFormatV2 can ONLY assignable to old version
+ */
+public class TestHiveHBaseTableInputFormatV2 {
+
+  @Test
+  public void testInstanceOfTestHiveHBaseTableInputFormatV2() {
```
Comment (aturoczy) [inline hbase-handler/src/test/org/apache/hadoop/hive/hbase/TestHiveHBaseTableInputFormatV2.java:33]: Not very detailed tests :)

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java (hunk 3 above)
Comment (alexdongli0829) [reply hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:88]: The connection is used by the RecordReader, so the close will be involved in the reader close() which at line 107

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java (hunk 4 above)
Comment (alexdongli0829) [reply hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:241]: ack

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java (hunk 2 above)
Comment (alexdongli0829) [reply hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:97]: ack

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java (hunk 1 above)
Comment (alexdongli0829) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:3510]: ack

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java (hunk 5 above)
Comment (alexdongli0829) [reply hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:252]: For this, actually the function is from https://github.com/apache/hive/blob/master/hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormat.java#L238. I would not change it in this PR. For me, I am not sure if its easy to detect the UDF, and maybe we can use a new thread to track if you think its a good optimization?

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java (hunk 6 above)
Comment (alexdongli0829) [reply hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:269]: I think this doAs is more about when use Auth like kerberos, then we need use UGI to get properly subject, right?

File: hbase-handler/src/test/org/apache/hadoop/hive/hbase/TestHiveHBaseTableInputFormatV2.java (hunk 7 above)
Comment (alexdongli0829) [reply hbase-handler/src/test/org/apache/hadoop/hive/hbase/TestHiveHBaseTableInputFormatV2.java:33]: For this class, I did not see many functions can be easily unit test, and I believe that`s why there was even no test for HiveHBaseTableInputFormat, do you have any suggestion on which function or how we can do more test?

**Comment by alexdongli0829** [comment]
@aturoczy could you help to check the update version?

**Comment by alexdongli0829** [comment]
Waiting for the review and merge

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 8]
```diff
@@ -3504,6 +3504,12 @@ public static enum ConfVars {
     HIVE_HBASE_GENERATE_HFILES("hive.hbase.generatehfiles", false,
         "True when HBaseStorageHandler should generate hfiles instead of operate against the online table."),
     HIVE_HBASE_SNAPSHOT_NAME("hive.hbase.snapshot.name", null, "The HBase table snapshot name to use."),
+
+    HIVE_HBASE_INPUTFORMAT_V2("hive.hbase.inputformat.v2", false, "If enabled, the new " +
+            "version (V2) of the input format, which inherits only the mapred version of InputFormat, will be " +
+            "utilized as the default input format for reading the Hbase table. " +
+            "Currently, the old version (HiveHBaseTableInputFormat) is used by default, " +
+            "which inherits both the mapred and mapreduce versions of InputFormat."),
```
Comment (ayushtkn) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:3512]: I think this config isn't right, we should just have ``hive.hbase.inputformat`` and specify ``1`` as default  and use ``2`` for v2, else if tmrw v3 or v4 pitches in we would be having 1 config for each

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+import java.util.Map;
+
+/**
+ * It`s the pair with {@link HiveHBaseTableInputFormat}, instead of directly extending TableInputFormatBase, using delegate
+ * to make the class ONLY inherit from mapred.*, which makes the hierarchy more clear and avoid downstream application
+ * like spark issue, ex https://github.com/apache/spark/pull/31302
+ */
+public class HiveHBaseTableInputFormatV2 implements InputFormat<ImmutableBytesWritable, ResultWritable> {
+
+  static final Logger LOG = LoggerFactory.getLogger(HiveHBaseTableInputFormatV2.class);
```
Comment (ayushtkn) [inline hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:69]: shouldn't be ``private``?

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+  private static final Object HBASE_TABLE_MONITOR = new Object();
+
+  private HiveHBaseTableInputFormatDelegate delegate = new HiveHBaseTableInputFormatDelegate();
```
Comment (ayushtkn) [inline hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:72]: should be final

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+            value.setResult(recordReader.getCurrentValue());
+          }
+        } catch (InterruptedException e) {
+          throw new IOException(e);
+        }
+
+        return next;
+      }
+    };
+  }
```
Comment (ayushtkn) [inline hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:153]: this is an exact dupe from ``HiveHBaseTableInputFormat``, avoid duplication and refactor

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+   * into corresponding restrictions on the HBase scan.  The
+   * filter should already be in a form which can be fully converted.
+   *
+   * @param jobConf configuration for the scan
+   *
+   * @param iKey 0-based offset of key column within Hive table
+   *
+   * @return converted table split if any
+   */
+  private Scan createFilterScan(JobConf jobConf, int iKey, int iTimestamp, boolean isKeyBinary) throws IOException {
```
Comment (ayushtkn) [inline hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:166]: This method is also a complete copy from v1

File: hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+              genericUDFOPEqualOrLessThanClass,
+              genericUDFOPLessThanClass,
+              genericUDFOPGreaterThanClass);
+    }
+
+    return analyzer;
+  }
+
+  @Override public InputSplit[] getSplits(final JobConf jobConf, final int numSplits) throws IOException {
+    synchronized (HBASE_TABLE_MONITOR) {
```
Comment (ayushtkn) [inline hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java:267]: this is also pretty much a copy from v1

**Review by ayushtkn — CHANGES_REQUESTED** [review]
I started reviewing, it has lot of duplication from v1, we should use some refactor strategy to avoid duplications, some util class or best to have some base class/interface where these duplicate methods can be defined and used in both classes.

Will check once this dupe stuff is sorted
